### PR TITLE
Fix ZMQ bug

### LINKED ISSUE #168: ZMQ error
Hi,

While running PyMeasure on a newer version of anaconda, I got an error which looks like it's coming from an update ZMQ had. In particular, the error occurs when a `worker` tries to use its `emit` method. The error is in the `self.publisher.send_serialized` function, which I'm guessing has changed in a newer version of ZMQ or something. A fix for now is to add a `TypeError` to the following `except` statement, but some functionality (which it seems I don't use) is probably being lost along the way. The full error message is at the bottom of this post, and the versions of everything that I'm using are below.

ZMQ version: 17.1.2
Python version: 3.6.6
Conda version: 4.5.11

```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Users\Joseph\Anaconda3\lib\site-packages\zmq\sugar\socket.py", line 427, in send_multipart
    memoryview(msg)
TypeError: memoryview: a bytes-like object is required, not 'int'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Joseph\Anaconda3\lib\threading.py", line 916, in _bootstrap_inner
    self.run()
  File "C:\Users\Joseph\Anaconda3\lib\site-packages\pymeasure-0.5.1-py3.6.egg\pymeasure\experiment\workers.py", line 161, in run
    self.update_status(Procedure.RUNNING)
  File "C:\Users\Joseph\Anaconda3\lib\site-packages\pymeasure-0.5.1-py3.6.egg\pymeasure\experiment\workers.py", line 115, in update_status
    self.emit('status', status)
  File "C:\Users\Joseph\Anaconda3\lib\site-packages\pymeasure-0.5.1-py3.6.egg\pymeasure\experiment\workers.py", line 95, in emit
    self.publisher.send_serialized((topic, record), serialize=cloudpickle.dumps)
  File "C:\Users\Joseph\Anaconda3\lib\site-packages\zmq\sugar\socket.py", line 512, in send_serialized
    return self.send_multipart(frames, flags=flags, copy=copy, **kwargs)
  File "C:\Users\Joseph\Anaconda3\lib\site-packages\zmq\sugar\socket.py", line 434, in send_multipart
    i, rmsg,
TypeError: Frame 0 (128) does not support the buffer interface.
```

### PR DESCRIPTION
**Pull request by bilderbuchi**
Closes #168.

Let's first see that we can trigger the problem.

### CODE REVIEW

**Comment by bilderbuchi** [comment]
OK, I have started a proposed fix, but I am running into weird issues. Locally (py3.7), if I run the new `test_zmq_serialisation` test, the test passes but then pytest hangs until I press Ctrl-C:
```
(base) bilderbuchi@osprey:~/Python/pymeasure$ pytest -rsx -v -k zmq
========================================================================= test session starts =========================================================================
platform linux -- Python 3.7.9, pytest-6.1.1, py-1.9.0, pluggy-0.13.1 -- /home/bilderbuchi/anaconda3/bin/python
cachedir: .pytest_cache
PyQt5 5.9.2 -- Qt runtime 5.9.7 -- Qt compiled 5.9.6
rootdir: /home/bilderbuchi/Python/pymeasure
plugins: qt-3.3.0, datadir-1.3.2.dev1+gf4f6a5f
collected 73 items / 72 deselected / 1 skipped                                                                                                                        

tests/experiment/test_workers.py::test_zmq_serialisation PASSED                                                                                                 [100%]

^CException ignored in: <function Context.__del__ at 0x7fa6a408f4d0>
Traceback (most recent call last):
  File "/home/bilderbuchi/anaconda3/lib/python3.7/site-packages/zmq/sugar/context.py", line 54, in __del__
    self.term()
  File "/home/bilderbuchi/anaconda3/lib/python3.7/site-packages/zmq/sugar/context.py", line 154, in term
    return super(Context, self).term()
  File "zmq/backend/cython/context.pyx", line 91, in zmq.backend.cython.context.Context.term
  File "zmq/backend/cython/checkrc.pxd", line 13, in zmq.backend.cython.checkrc._check_rc
KeyboardInterrupt

======================================================================= short test summary info =======================================================================
SKIPPED [1] tests/instruments/agilent/test_agilent34450A.py:30: Only work with connected hardware
============================================================= 1 passed, 1 skipped, 72 deselected in 4.55s =============================================================
```
In CI, this fails even

**Comment by bilderbuchi** [comment]
By now the indications are becoming more that this is a problem with not setting the ZQM Linger setting: https://stackoverflow.com/questions/44273941/python-script-not-terminating-after-timeout-in-zmq-recv https://stackoverflow.com/questions/7939977/zeromq-with-python-hangs-if-connecting-to-invalid-socket
I still have to check if this can be solved by a cleaner shutdown procedure, i.e. making sure to consume all messages before shutdown, or just setting a `linger` value and being done with it.

**Comment by bilderbuchi** [comment]
hey @CasperSchippers @Kassikakk @jmittelstaedt: I have fixed the ZMQ problems reported in #168. Could you maybe test if this PR works in your real workflow? 
I have added some tests that are passing (on all platforms now), but want to make sure that this also covers a real-life scenario.

File: pymeasure/experiment/workers.py [hunk 1]
```diff
@@ -76,8 +76,26 @@ def __init__(self, results, log_queue=None, log_level=logging.INFO, port=None):
...
+            try:
+                self.context = zmq.Context()
+                log.debug("Worker ZMQ Context: %r" % self.context)
+                self.publisher = self.context.socket(zmq.PUB)
+                self.publisher.bind('tcp://*:%d' % self.port)
+                log.info("Worker connected to tcp://*:%d" % self.port)
+                time.sleep(0.3)  # wait so that the socket will be ready before starting to emit messages
+            except Exception:
+                log.exception("couldn't connect to ZMQ context")
+                # FIXME: this happily continues even if connecting to the context fails!
```
Comment (CasperSchippers) [inline pymeasure/experiment/workers.py:98]: What about this remark? Is this a problem (it's not completely clear to me what happens if connecting fails)

**Review by CasperSchippers — COMMENTED** [review]
Awesome! Well done @bilderbuchi!
I tested it in various environments and as far as I can see it works like a charm.

File: tests/experiment/test_workers.py [hunk 2]
```diff
@@ -82,3 +85,55 @@ def test_worker_closes_file_after_finishing():
 
     # Test if the file has been properly closed by removing the file
     os.remove(file)
+
+
+@pytest.mark.skipif(not importlib.util.find_spec('cloudpickle'),
```
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply tests/experiment/test_workers.py:90]: Indeed!

File: pymeasure/experiment/workers.py (hunk 1 above)
Comment (bilderbuchi) [reply pymeasure/experiment/workers.py]: I wanted to postpone that, but indeed it's better to do it right away - I let this fall back to proper values (as if zmq were not installed)

**Comment by bilderbuchi** [comment]
Thanks for your feedback, I'll merge this, then. :-)

**Pull request merged**